source/opt/unify_const_pass.cpp

```diff
@@ -0,0 +1,185 @@
+  ResultIdTrieNode* GetOrCreateTrieNodeFor(uint32_t key) {
+    auto iter = next_.find(key);
+    if (iter == next_.end()) {
+      // insert new entry
+      return next_.insert(std::make_pair(key, MakeUnique<ResultIdTrieNode>()))
+          .first->second.get();
+    }
+    return iter->second.get();
+  }
+
+ private:
+  // The result id stored in this node. 0 means this node is empty.
+  uint32_t result_id_;
+  // The mapping from the keys to the child nodes of this node.
+  TrieNodeMap next_;
+};
+
+// The trie that stores a bunch of result ids and performs equivalent result id
+// looking up with a given instruction.
+class ResultIdTrie {
+ public:
+  ResultIdTrie() : root_(new ResultIdTrieNode) {}
+
+  // For a given instruction, extracts its opcode, type id and operand words
+  // as an array of keys, looks up the trie to find an result id which is
+  // stored with the same opcode, type id and operand words. If none of such
+  // result id is found, creates a trie node with those keys, stores the
+  // instruction's result id and returns that result id. If a existing result
+  // id is found, returns the existing result id.
+  uint32_t LookupEquivalentResultFor(const ir::Instruction& inst) {
+    auto keys = GetLookUpKeys(inst);
+    auto* node = root_.get();
+    for (uint32_t key : keys) {
+      node = node->GetOrCreateTrieNodeFor(key);
+    }
+    if (node->result_id() == 0) {
+      node->SetResultId(inst.result_id());
+    }
+    return node->result_id();
+  }
+
+ private:
+  // Extracts the opcode, result type id and operand words from an instruction.
+  // Returns a vector of those values in 32-bit unsigned integers.
+  std::vector<uint32_t> GetLookUpKeys(const ir::Instruction& inst) {
+    std::vector<uint32_t> keys;
+    // Need to use the opcode, otherwise there might be a conflict with the
+    // following case when <op>'s binary value equals xx's id:
+    //  OpSpecConstantOp tt <op> yy zz
+    //  OpSpecConstantComposite tt xx yy zz;
+    keys.push_back(static_cast<uint32_t>(inst.opcode()));
+    for (uint32_t i = 0; i < inst.NumOperands(); i++) {
+      if (inst.GetOperand(i).type == SPV_OPERAND_TYPE_RESULT_ID) continue;
+      keys.insert(keys.end(), inst.GetOperand(i).words.cbegin(),
+                  inst.GetOperand(i).words.cend());
+    }
+    return keys;
+  }
+
+  std::unique_ptr<ResultIdTrieNode> root_;  // The root node of the trie.
+};
+}  // anonymous namespace
+
+bool UnifyConstantPass::Process(ir::Module* module) {
+  bool modified = false;
+  ResultIdTrie defined_constants;
+  std::unordered_set<ir::Instruction*> duplicated_cosnts;
+  analysis::DefUseManager def_use_mgr(module);
+
+  for (ir::Instruction& inst : module->types_values()) {
+    // Do not handle the instruction when there are decorations upon either
+    // the result id or the type id.
+    if (def_use_mgr.GetAnnotations(inst.result_id()).size() != 0 ||
+        def_use_mgr.GetAnnotations(inst.type_id()).size() != 0) {
```

Comment:
If type id is the same, the decorations on types should be the same. So it's fine to have decorations on types since we use the type id as a part of the key.